staging/kos/pkg/controllers/connectionagent/connectionagent.go

```diff
@@ -0,0 +1,1091 @@
+	vnState := ca.vniToVnState[vni]
+	if vnState == nil {
+		return nil, nil
+	}
+	return vnState.remoteAttsInformer, vnState.remoteAttsInformerStopCh
+}
+
+// Return a string representing a field selector that matches NetworkAttachments
+// that run on the local node and have a virtual IP.
+func (ca *ConnectionAgent) localAttWithAnIPSelector() string {
+	// localAttSelector expresses the constraint that the NetworkAttachment runs on this node.
+	localAttSelector := attNodeFieldName + equal + ca.localNodeName
+
+	// Express the constraint that the NetworkAttachment has a virtual IP by saying
+	// that the field containig the virtual IP must not be equal to the empty string.
+	attWithAnIPSelector := attIPFieldName + notEqual
+
+	// Build a selector which is a logical AND between
+	// attWithAnIPSelectorString and localAttSelectorString.
+	allSelectors := []string{localAttSelector, attWithAnIPSelector}
+	return strings.Join(allSelectors, ",")
+}
+
+// Return a string representing a field selector that matches NetworkAttachments
+// that run on a remote node on the Virtual Network identified by the given VNI and
+// have a virtual IP and the host IP field set.
+func (ca *ConnectionAgent) remoteAttInVNWithVirtualIPHostIPAndIfcSelector(vni uint32) string {
+	// remoteAttSelector expresses the constraint that the NetworkAttachment runs on a remote node.
+	remoteAttSelector := attNodeFieldName + notEqual + ca.localNodeName
+
+	// hostIPIsNotLocalSelector expresses the constraint that the NetworkAttachment status.hostIP is not equal
+	// to that of the current node. Without this selector, an update to the spec.Node field of a NetworkAttachment
+	// could lead to a creation notification for the attachment in a remote attachments cache, even if the attachment
+	// still has the host IP of the current node (status.hostIP is set with an update by the connection agent on the
+	// node of the attachment). This could result in the creation of a remote interface with the host IP of the local
+	// node.
+	hostIPIsNotLocalSelector := attHostIPFieldName + notEqual + ca.hostIP.String()
+
+	// attWithAnIPSelector, attWithHostIPSelector and attWithIfcSelector express the constraints that the
+	// NetworkAttachment has the fields storing virtual IP, host IP and ifc name sets, by saying that such
+	// fields must not be equal to the empty string.
+	attWithAnIPSelector := attIPFieldName + notEqual
+	attWithHostIPSelector := attHostIPFieldName + notEqual
+	attWithIfcSelector := attIfcFieldName + notEqual
+
+	// attInSpecificVNSelctor expresses the constraint that the NetworkAttachment is in the Virtual
+	// Network identified by vni.
+	attInSpecificVNSelctor := attVNIFieldName + equal + fmt.Sprint(vni)
+
+	// Build and return a selector which is a logical AND between all the selectors defined above.
+	allSelectors := []string{remoteAttSelector,
+		hostIPIsNotLocalSelector,
+		attWithAnIPSelector,
+		attWithHostIPSelector,
+		attWithIfcSelector,
+		attInSpecificVNSelctor}
```

Comment:
I suggest listing this first because it is the most discriminating, which I expect would make this condition faster to evaluate on average.

Reply:
I'm not sure I understand. Are you talking about moving attInSpecificVNSelector as the first one because it's more discriminating? I think that when the string is parsed into a selector the individual selectors are sorted, so we have no control over the order with which they are applied: https://github.com/kubernetes/apimachinery/blob/master/pkg/fields/selector.go#L430

Reply:
Dang!